### PR TITLE
Cross-platform traceStorage

### PR DESCRIPTION
**Pull request by dellaert**
Because variable arrays is a C99 feature, and we were doing AD trace storage differently on Windows/Linux, I unified both platforms with some help from ChatGPT. In doing this, I can see an even better way, but that will have to come later.

Update: this PR also fixes the memory overruns that were occurring in several paths of the AD code, because of faulty pointer arithmetic, *and* reverts back to 32 bit alignment if not on Windows, because Eigen can choose that alignment on some platforms. The latter needs a solution where we can find out at compile time what Eigen's chosen alignment is.

### CODE REVIEW

File: tests/testExpressionFactor.cpp [hunk 1]
```diff
@@ -434,17 +430,17 @@ TEST(ExpressionFactor, composeTernary) {
   // Check unwhitenedError
   std::vector<Matrix> H(3);
   Vector actual = f.unwhitenedError(values, H);
-  EXPECT_LONGS_EQUAL(3, H.size());
-  EXPECT(assert_equal(I_3x3, H[0],1e-9));
-  EXPECT(assert_equal(I_3x3, H[1],1e-9));
-  EXPECT(assert_equal(I_3x3, H[2],1e-9));
+  EXPECT_LONGS_EQUAL(3, H.size())
+  EXPECT(assert_equal(I_3x3, H[0],1e-9))
```
Comment (ankur-rc) [inline tests/testExpressionFactor.cpp:434]: Formatting: missing a space after the comma

File: tests/testExpressionFactor.cpp [hunk 2]
```diff
@@ -434,17 +430,17 @@ TEST(ExpressionFactor, composeTernary) {
...
+  EXPECT(assert_equal(I_3x3, H[1],1e-9))
+  EXPECT(assert_equal(I_3x3, H[2],1e-9))
 
   // Check linearization
   JacobianFactor expected(1, I_3x3, 2, I_3x3, 3, I_3x3, Z_3x1);
   std::shared_ptr<GaussianFactor> gf = f.linearize(values);
   std::shared_ptr<JacobianFactor> jf = //
       std::dynamic_pointer_cast<JacobianFactor>(gf);
-  EXPECT(assert_equal(expected, *jf,1e-9));
+  EXPECT(assert_equal(expected, *jf,1e-9))
```
Comment (ankur-rc) [inline tests/testExpressionFactor.cpp:443]: Formatting: missing a space after comma

File: tests/testExpressionFactor.cpp [hunk 3]
```diff
@@ -333,15 +329,15 @@ TEST(ExpressionFactor, Compose1) {
   // Check unwhitenedError
   std::vector<Matrix> H(2);
   Vector actual = f.unwhitenedError(values, H);
-  EXPECT(assert_equal(I_3x3, H[0],1e-9));
-  EXPECT(assert_equal(I_3x3, H[1],1e-9));
+  EXPECT(assert_equal(I_3x3, H[0],1e-9))
```
Comment (ankur-rc) [inline tests/testExpressionFactor.cpp:332]: Formatting: missing a space after comma

File: tests/testExpressionFactor.cpp [hunk 4]
```diff
@@ -333,15 +329,15 @@ TEST(ExpressionFactor, Compose1) {
...
+  EXPECT(assert_equal(I_3x3, H[1],1e-9))
 
   // Check linearization
   JacobianFactor expected(1, I_3x3, 2, I_3x3, Z_3x1);
   std::shared_ptr<GaussianFactor> gf = f.linearize(values);
   std::shared_ptr<JacobianFactor> jf = //
       std::dynamic_pointer_cast<JacobianFactor>(gf);
-  EXPECT(assert_equal(expected, *jf,1e-9));
+  EXPECT(assert_equal(expected, *jf,1e-9))
```
Comment (ankur-rc) [inline tests/testExpressionFactor.cpp:340]: Formatting: missing a space after comma

File: gtsam/nonlinear/internal/ExecutionTrace.h [hunk 5]
```diff
@@ -37,7 +38,13 @@ template<int T> struct CallRecord;
 /// Storage type for the execution trace.
 /// It enforces the proper alignment in a portable way.
 /// Provide a traceSize() sized array of this type to traceExecution as traceStorage.
-static const unsigned TraceAlignment = 16; // 16 bytes is the default alignment used by Eigen.
+#ifdef _MSC_VER
```
Comment (ankur-rc) [inline gtsam/nonlinear/internal/ExecutionTrace.h:41]: Defining `EIGEN_MAX_ALIGN_BYTES` maybe a solution.
Reference: https://eigen.tuxfamily.org/dox/TopicPreprocessorDirectives.html

**Review by ankur-rc — APPROVED** [review]

File: tests/testExpressionFactor.cpp (hunk 1 above)
Comment (dellaert) [reply tests/testExpressionFactor.cpp:434]: Thanks ! Note I won't change as orthogonal to the PR - there are *many* formatting issues with al tests.

File: gtsam/nonlinear/internal/ExpressionNode.h [hunk 6]
```diff
@@ -625,7 +627,7 @@ class ScalarMultiplyNode : public ExpressionNode<T> {
 
   /// Construct an execution trace for reverse AD
   T traceExecution(const Values& values, ExecutionTrace<T>& trace,
-                           ExecutionTraceStorage* ptr) const override {
+                           char* ptr) const override {
     assert(reinterpret_cast<size_t>(ptr) % TraceAlignment == 0);
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/nonlinear/internal/ExpressionNode.h:631]: Correct, as intended.

File: gtsam/nonlinear/internal/ExecutionTrace.h (hunk 5 above)
Comment (dellaert) [reply gtsam/nonlinear/internal/ExecutionTrace.h:41]: Thanks ! I'll read that...

**Pull request merged**